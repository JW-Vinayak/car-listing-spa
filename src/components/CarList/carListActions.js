import { getCarList } from "../../utils/api";
import { formQueryString } from "../../utils/utility";

export const FETCH_CARS_BEGIN = "FETCH_CARS_BEGIN";
export const FETCH_CARS_SUCCESS = "FETCH_CARS_SUCCESS";
export const FETCH_CARS_FAILURE = "FETCH_CARS_FAILURE";
export const SET_RECORDS_INFO = "SET_RECORDS_INFO";

export const fetchCarsBegin = () => ({
  type: FETCH_CARS_BEGIN
});

export const fetchCarsSuccess = cars => ({
  type: FETCH_CARS_SUCCESS,
  payload: { cars }
});

export const fetchCarsFailure = error => ({
  type: FETCH_CARS_FAILURE,
  payload: { error }
});

export const setRecordInfo = recordsInfo => ({
  type: SET_RECORDS_INFO,
  payload: { recordsInfo }
});

export function fetchCars() {
  return (dispatch, getState) => {
    let params = {},
      filters = getState().filters,
      pagination = getState().pagination;
    if (filters.color) {
      params.color = filters.color;
    }
    if (filters.manufacturer) {
      params.manufacturer = filters.manufacturer;
    }
    if (pagination.currentPage) {
      params.page = pagination.currentPage;
    }

    let queryString = formQueryString(params);

    console.log("will fire api with", params, queryString);
    dispatch(fetchCarsBegin());
    console.log("sent request to fetch data");

    getCarList(queryString)
      .then(json => {
        console.log("data is fetched", json);
        dispatch(fetchCarsSuccess(json.cars));
        dispatch(
          setRecordInfo({
            totalPages: json.totalPageCount,
            totalRecords: json.totalCarsCount
          })
        );
        console.log("data after dispatch", getState());
        return json.cars;
      })
      .catch(error => dispatch(fetchCarsFailure(error)));
  };
}
