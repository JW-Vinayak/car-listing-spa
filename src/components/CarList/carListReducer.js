import {
  FETCH_CARS_BEGIN,
  FETCH_CARS_SUCCESS,
  FETCH_CARS_FAILURE,
  SET_RECORDS_INFO
} from "./carListActions";

const initialState = {
  list: [],
  loading: false,
  error: null,
  totalPages: 0,
  totalRecords: 0
};

export default function carListReducer(state = initialState, action) {
  switch (action.type) {
    case FETCH_CARS_BEGIN:
      return {
        ...state,
        loading: true,
        error: null
      };

    case FETCH_CARS_SUCCESS:
      return {
        ...state,
        loading: false,
        list: action.payload.cars
      };

    case FETCH_CARS_FAILURE:
      return {
        ...state,
        loading: false,
        error: action.payload.error,
        list: []
      };

    case SET_RECORDS_INFO:
      console.log('SET_TOTAL_PAGES', action)
      return {
        ...state,
        totalPages: action.payload.recordsInfo.totalPages,
        totalRecords: action.payload.recordsInfo.totalRecords
      };

    default:
      return state;
  }
}
