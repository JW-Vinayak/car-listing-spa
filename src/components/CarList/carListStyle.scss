.car-list {
  .car-image {
    height: 30px;
    width: 30px;
  }
}
