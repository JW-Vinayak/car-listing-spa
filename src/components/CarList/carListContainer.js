import { connect } from "react-redux";
import { fetchCars } from "./carListActions";
import CarList from "./carListView";

const mapStateToProps = state => ({
  cars: state.cars.list,
  loading: state.cars.loading,
  error: state.cars.error,
  filters: state.filters,
  pagination: state.pagination,
  mileage: state.sorter.mileage
});

const mapDispatchToProps = dispatch => ({
  fetchCars: () => dispatch(fetchCars())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CarList);
