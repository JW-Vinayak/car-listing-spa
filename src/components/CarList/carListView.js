import React, { useEffect } from "react";
import PropTypes from "prop-types";
import "./carListStyle.scss";
import CarItem from "./CarItem/CarItemView";
import { sortCars } from "../../utils/utility";

export default function CarList(props) {
  let { filters, fetchCars, pagination, mileage } = props;
  let modifiedCarList = [];
  useEffect(() => {
    fetchCars();
  }, [filters, pagination, mileage]);

  const { error, loading, cars } = props;

  if (error) {
    return (
      <div class="error error-box">
        Couldn't load car list from server at this moment, please try again
        later.
      </div>
    );
  }

  if (loading) {
    return (
      <div className="car-list">
        {[...Array(10)].map((e, i) => (
          <CarItem key={i} car={{}} />
        ))}
      </div>
    );
  }
  return (
    <div className="car-list">
      {cars &&
        cars.map(car => <CarItem key={car.stockNumber} car={car} />)}
    </div>
  );
}

CarList.propTypes = {
  filters: PropTypes.object.isRequired,
  fetchCars: PropTypes.func.isRequired,
  pagination: PropTypes.object.isRequired,
  mileage: PropTypes.string.isRequired,
  error: PropTypes.object,
  loading: PropTypes.bool.isRequired,
  cars: PropTypes.array.isRequired
};
