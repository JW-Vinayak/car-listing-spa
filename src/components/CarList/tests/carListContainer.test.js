import React from "react";
import thunk from "redux-thunk";
import CarList from "../carListContainer";
import configureMockStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

const middlewares = [thunk];

const mockStore = configureMockStore(middlewares);

let store = mockStore({
  cars: {
    list: [],
    loading: false,
    error: false
  },
  filters: {},
  pagination: {},
  sorter: {}
});

describe("Car list component", () => {
  it("shallow renders without crashing", () => {
    const tree = renderer
      .create(
        <Router>
          <CarList store={store} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders two car items", () => {
    store = mockStore({
      cars: {
        list: [
          {
            stockNumber: 10019,
            manufacturerName: "Mercedes-Benz",
            modelName: "GLC-Klasse",
            color: "blue",
            mileage: { number: 181526, unit: "km" },
            fuelType: "Petrol"
          },
          {
            stockNumber: 10019,
            manufacturerName: "BMW",
            modelName: "i8",
            color: "black",
            mileage: { number: 12112, unit: "km" },
            fuelType: "Petrol"
          }
        ],
        loading: false,
        error: false
      },
      filters: { color: "", manufacturer: "" },
      pagination: { currentPage: 1 },
      sorter: { mileage: "NO_MILEAGE_SORTING" }
    });
    const tree = renderer
      .create(
        <Router>
          <CarList store={store} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders loading blocks if loading flag is true", () => {
    store = mockStore({
      cars: {
        list: [],
        loading: true,
        error: false
      },
      filters: {},
      pagination: {},
      sorter: {}
    });
    const tree = renderer
      .create(
        <Router>
          <CarList store={store} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders sorted car item due to specified filter", () => {
    store = mockStore({
      cars: {
        list: [
          {
            stockNumber: 10019,
            manufacturerName: "Mercedes-Benz",
            modelName: "GLC-Klasse",
            color: "blue",
            mileage: { number: 181526, unit: "km" },
            fuelType: "Petrol"
          },
          {
            stockNumber: 10019,
            manufacturerName: "BMW",
            modelName: "i8",
            color: "black",
            mileage: { number: 12112, unit: "km" },
            fuelType: "Petrol"
          }
        ],
        loading: false,
        error: false
      },
      filters: { color: "blue", manufacturer: "" },
      pagination: { currentPage: 1 },
      sorter: { mileage: "ASCENDING_MILEAGE_SORTING" }
    });
    const tree = renderer
      .create(
        <Router>
          <CarList store={store} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
