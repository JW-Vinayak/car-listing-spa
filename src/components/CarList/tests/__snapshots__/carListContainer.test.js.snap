// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Car list component renders loading blocks if loading flag is true 1`] = `
<div
  className="car-list"
>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
  <div
    className="car-item"
  >
    <div
      className="empty-box"
    />
    <div
      className="car-info"
    >
      <div
        className="no-car-name"
      />
      <div
        className="no-car-details"
      />
      <div
        className="no-view-details-link"
      />
    </div>
  </div>
</div>
`;

exports[`Car list component renders sorted car item due to specified filter 1`] = `
<div
  className="car-list"
>
  <div
    className="car-item"
  >
    <img
      alt="a car"
      className="car-image"
    />
    <div
      className="car-info"
    >
      <h3
        className="car-name"
      >
        i8
      </h3>
      <div
        className="car-details"
      >
        <span>
          Stock # 
          10019
           - 
        </span>
        <span>
          12112
           
          KM
           -
           
        </span>
        <span>
          Petrol
           - 
        </span>
        <span>
          black
        </span>
      </div>
      <a
        className="view-details-link"
        href="/car/10019"
        onClick={[Function]}
      >
        View Details
      </a>
    </div>
  </div>
  <div
    className="car-item"
  >
    <img
      alt="a car"
      className="car-image"
    />
    <div
      className="car-info"
    >
      <h3
        className="car-name"
      >
        GLC-Klasse
      </h3>
      <div
        className="car-details"
      >
        <span>
          Stock # 
          10019
           - 
        </span>
        <span>
          181526
           
          KM
           -
           
        </span>
        <span>
          Petrol
           - 
        </span>
        <span>
          blue
        </span>
      </div>
      <a
        className="view-details-link"
        href="/car/10019"
        onClick={[Function]}
      >
        View Details
      </a>
    </div>
  </div>
</div>
`;

exports[`Car list component renders two car items 1`] = `
<div
  className="car-list"
>
  <div
    className="car-item"
  >
    <img
      alt="a car"
      className="car-image"
    />
    <div
      className="car-info"
    >
      <h3
        className="car-name"
      >
        GLC-Klasse
      </h3>
      <div
        className="car-details"
      >
        <span>
          Stock # 
          10019
           - 
        </span>
        <span>
          181526
           
          KM
           -
           
        </span>
        <span>
          Petrol
           - 
        </span>
        <span>
          blue
        </span>
      </div>
      <a
        className="view-details-link"
        href="/car/10019"
        onClick={[Function]}
      >
        View Details
      </a>
    </div>
  </div>
  <div
    className="car-item"
  >
    <img
      alt="a car"
      className="car-image"
    />
    <div
      className="car-info"
    >
      <h3
        className="car-name"
      >
        i8
      </h3>
      <div
        className="car-details"
      >
        <span>
          Stock # 
          10019
           - 
        </span>
        <span>
          12112
           
          KM
           -
           
        </span>
        <span>
          Petrol
           - 
        </span>
        <span>
          black
        </span>
      </div>
      <a
        className="view-details-link"
        href="/car/10019"
        onClick={[Function]}
      >
        View Details
      </a>
    </div>
  </div>
</div>
`;

exports[`Car list component shallow renders without crashing 1`] = `
<div
  className="car-list"
/>
`;
